src/grelu/data/dataset.py

```diff
@@ -1011,3 +1012,250 @@ def __getitem__(self, idx: int, return_compressed=False) -> Tensor:
+        fixed_pattern: A subsequence to insert in the center of each background sequence.
+        moving_pattern: A subsequence to insert into the background sequences at
+            different distances from `fixed_motif`.
+        stride: Number of bases by which to shift the moving pattern.
+        genome: The name of the genome from which to read sequences. This
+            is only needed if genomic intervals are supplied in `seqs`.
+        n_shuffles: Number of times to shuffle each sequence in `seqs`, to
+            generate a background distribution.
+        seed: Seed for random number generator
+    """
+
+    def __init__(
+        self,
+        seqs: Union[str, Sequence, pd.DataFrame, np.ndarray],
+        fixed_pattern: str,
+        moving_pattern: str,
+        stride: int = 1,
+        genome: Optional[str] = None,
+        n_shuffles: int = 1,
+        rc: bool = False,
+        seed: int = 0,
+    ) -> None:
+        super().__init__()
+
+        # Save params
+        self.stride = stride
+        self.genome = genome
+        self.n_shuffles = n_shuffles
+        self.seed = seed
+        self.rc = rc
+
+        # Ingest sequences
+        self._load_seqs(seqs)
+
+        # Ingest patterns
+        self._load_patterns(fixed_pattern, moving_pattern)
+
+        # Calculate positions
+        self._set_positions()
+
+        # Create augmenter
+        self.augmenter = Augmenter(
+            rc=self.rc,
+            seq_len=self.seq_len,
+            mode="serial",
+        )
+        self.n_augmented = len(self.augmenter)
+
+        # Initial state
+        self.bg = None
+        self.curr_seq_idx = None
+
+    def _load_seqs(self, seqs: Union[pd.DataFrame, List[str], np.ndarray]) -> None:
+        """
+        Make the background sequences
+        """
+        self.seqs = convert_input_type(
+            seqs, "indices", genome=self.genome, add_batch_axis=True
+        )
+        self.n_seqs = self.seqs.shape[0]
+        self.seq_len = self.seqs.shape[1]
+
+    def _load_patterns(self, fixed_pattern: str, moving_pattern: str) -> None:
+        self.fixed_pattern = strings_to_indices(fixed_pattern)
+        self.moving_pattern = strings_to_indices(moving_pattern)
+        self.fixed_pattern_len = len(self.fixed_pattern)
+        self.moving_pattern_len = len(self.moving_pattern)
+
+    def _set_positions(self) -> None:
+
+        # Coordinates of the fixed pattern
+        self.fixed_pattern_start = int(
+            np.floor(self.seq_len / 2 - self.fixed_pattern_len / 2)
+        )
+        self.fixed_pattern_end = self.fixed_pattern_start + self.fixed_pattern_len
+
+        # Coordinates of the moving pattern
+        max_pos = self.seq_len - self.moving_pattern_len + 1
+        excl_start = self.fixed_pattern_start - self.moving_pattern_len + 1
+        excl = range(excl_start, self.fixed_pattern_end)
+
+        positions = [x for x in range(0, max_pos, self.stride) if x not in excl]
```

Comment:
This logic is shared with Tiling Shuffle Dataset. Should it be abstracted out? Do you see it being used again?

Reply:
Good idea, I will break it out